File: md_studio/utils/date_utils.py
from datetime import datetime
from typing import Union

def format_relative_date(date: Union[str, datetime]) -> str:
    now = datetime.now()
    then = datetime.fromisoformat(date) if isinstance(date, str) else date
    diff_seconds = int((now - then).total_seconds())
    
    if diff_seconds < 60:
        return "just now"
    
    diff_minutes = diff_seconds // 60
    if diff_minutes < 60:
        return f"{diff_minutes} {'minute' if diff_minutes == 1 else 'minutes'} ago"
    
    diff_hours = diff_minutes // 60
    if diff_hours < 24:
        return f"{diff_hours} {'hour' if diff_hours == 1 else 'hours'} ago"
    
    diff_days = diff_hours // 24
    if diff_days < 7:
        return f"{diff_days} {'day' if diff_days == 1 else 'days'} ago"
    
    if diff_days < 30:
        weeks = diff_days // 7
        return f"{weeks} {'week' if weeks == 1 else 'weeks'} ago"
    
    return then.strftime("%b %d, %Y")

def calculate_reading_time(text: str) -> int:
    words_per_minute = 200
    words = len([w for w in text.strip().split() if w])
    minutes = (words + words_per_minute - 1) // words_per_minute
    return max(1, minutes)

File: ui/lib/date-utils.ts
/**
 * Format a date as relative time (e.g., "2 hours ago")
 * Falls back to readable format for older dates
 */
export function formatRelativeDate(date: string | Date): string {
  const now = new Date();
  const then = typeof date === "string" ? new Date(date) : date;
  const diffInSeconds = Math.floor((now.getTime() - then.getTime()) / 1000);

  if (diffInSeconds < 60) {
    return "just now";
  }

  const diffInMinutes = Math.floor(diffInSeconds / 60);
  if (diffInMinutes < 60) {
    return `${diffInMinutes} ${diffInMinutes === 1 ? "minute" : "minutes"} ago`;
  }

  const diffInHours = Math.floor(diffInMinutes / 60);
  if (diffInHours < 24) {
    return `${diffInHours} ${diffInHours === 1 ? "hour" : "hours"} ago`;
  }

  const diffInDays = Math.floor(diffInHours / 24);
  if (diffInDays < 7) {
    return `${diffInDays} ${diffInDays === 1 ? "day" : "days"} ago`;
  }

  if (diffInDays < 30) {
    const weeks = Math.floor(diffInDays / 7);
    return `${weeks} ${weeks === 1 ? "week" : "weeks"} ago`;
  }

  // For older dates, use readable format
  return then.toLocaleDateString("en-US", {
    year: "numeric",
    month: "short",
    day: "numeric",
  });
}

/**
 * Calculate estimated reading time based on word count
 * Assumes 200 words per minute average reading speed
 */
export function calculateReadingTime(text: string): number {
  const wordsPerMinute = 200;
  const words = text.trim().split(/\s+/).filter(Boolean).length;
  const minutes = Math.ceil(words / wordsPerMinute);
  return Math.max(1, minutes); // Minimum 1 minute
}

/**
 * Count words in text
 */
export function countWords(text: string): number {
  return text.trim().split(/\s+/).filter(Boolean).length;
}
